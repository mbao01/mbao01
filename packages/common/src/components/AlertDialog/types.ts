import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";

export type AlertDialogProps = React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Root>;

export type AlertDialogOverlayProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Overlay
> &
  VariantProps<typeof getDialogOverlayClasses>;

export type AlertDialogContentProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Content
> &
  VariantProps<typeof getDialogContentClasses>;

export type AlertDialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type AlertDialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type AlertDialogTitleProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Title
> &
  VariantProps<typeof getDialogTitleClasses>;

export type AlertDialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;

export type AlertDialogActionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;

export type AlertDialogCancelProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;
