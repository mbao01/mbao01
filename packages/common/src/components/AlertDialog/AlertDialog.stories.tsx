import type { Meta, StoryObj } from "@storybook/react-vite";
import { Button } from "../Button";
import { AlertDialog } from "./AlertDialog";

const withDialog = () => {
  return (
    <AlertDialog>
      <AlertDialog.Trigger asChild>
        <Button outline>Update profile</Button>
      </AlertDialog.Trigger>
      <AlertDialog.Content className="sm:max-w-[425px]">
        <AlertDialog.Header>
          <AlertDialog.Title>Are you sure?</AlertDialog.Title>
          <AlertDialog.Description>
            You are making changes to your profile here. Do you want to continue?
          </AlertDialog.Description>
        </AlertDialog.Header>
        <AlertDialog.Footer>
          <AlertDialog.Cancel variant="link">Cancel</AlertDialog.Cancel>
          <AlertDialog.Action>Continue</AlertDialog.Action>
        </AlertDialog.Footer>
      </AlertDialog.Content>
    </AlertDialog>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: "Components/AlertDialog",
  component: AlertDialog,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withDialog],
} satisfies Meta<typeof AlertDialog>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};
