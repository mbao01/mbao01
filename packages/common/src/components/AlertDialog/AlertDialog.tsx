"use client";

import * as React from "react";
import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import type {
  AlertDialogActionProps,
  AlertDialogCancelProps,
  AlertDialogContentProps,
  AlertDialogDescriptionProps,
  AlertDialogFooterProps,
  AlertDialogHeaderProps,
  AlertDialogOverlayProps,
  AlertDialogProps,
  AlertDialogTitleProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";
import { Button } from "../Button";

const AlertDialog = (props: AlertDialogProps) => (
  <AlertDialogPrimitive.Root {...props} />
);

const AlertDialogTrigger = AlertDialogPrimitive.Trigger;

const AlertDialogPortal = AlertDialogPrimitive.Portal;

const AlertDialogOverlay = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Overlay>,
  AlertDialogOverlayProps
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Overlay
    className={cn(getDialogOverlayClasses(), className)}
    {...props}
    ref={ref}
  />
));
AlertDialogOverlay.displayName = AlertDialogPrimitive.Overlay.displayName;

const AlertDialogContent = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Content>,
  AlertDialogContentProps
>(({ className, ...props }, ref) => (
  <AlertDialogPortal>
    <AlertDialogOverlay />
    <AlertDialogPrimitive.Content
      ref={ref}
      className={cn(getDialogContentClasses(), className)}
      {...props}
    />
  </AlertDialogPortal>
));
AlertDialogContent.displayName = AlertDialogPrimitive.Content.displayName;

const AlertDialogHeader = ({ className, ...props }: AlertDialogHeaderProps) => (
  <div className={cn(getDialogHeaderClasses(), className)} {...props} />
);
AlertDialogHeader.displayName = "AlertDialogHeader";

const AlertDialogFooter = ({ className, ...props }: AlertDialogFooterProps) => (
  <div className={cn(getDialogFooterClasses(), className)} {...props} />
);
AlertDialogFooter.displayName = "AlertDialogFooter";

const AlertDialogTitle = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Title>,
  AlertDialogTitleProps
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Title
    ref={ref}
    className={cn(getDialogTitleClasses(), className)}
    {...props}
  />
));
AlertDialogTitle.displayName = AlertDialogPrimitive.Title.displayName;

const AlertDialogDescription = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Description>,
  AlertDialogDescriptionProps
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Description
    ref={ref}
    className={cn(getDialogDescriptionClasses(), className)}
    {...props}
  />
));
AlertDialogDescription.displayName =
  AlertDialogPrimitive.Description.displayName;

const AlertDialogAction = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Action>,
  AlertDialogActionProps
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Action ref={ref} asChild>
    <Button className={cn(className)} {...props} />
  </AlertDialogPrimitive.Action>
));
AlertDialogAction.displayName = AlertDialogPrimitive.Action.displayName;

const AlertDialogCancel = React.forwardRef<
  React.ElementRef<typeof AlertDialogPrimitive.Cancel>,
  AlertDialogCancelProps
>(({ className, ...props }, ref) => (
  <AlertDialogPrimitive.Cancel ref={ref} asChild>
    <Button className={cn("mt-2 sm:mt-0", className)} {...props} />
  </AlertDialogPrimitive.Cancel>
));
AlertDialogCancel.displayName = AlertDialogPrimitive.Cancel.displayName;

AlertDialog.Portal = AlertDialogPortal;
AlertDialog.Overlay = AlertDialogOverlay;
AlertDialog.Trigger = AlertDialogTrigger;
AlertDialog.Content = AlertDialogContent;
AlertDialog.Header = AlertDialogHeader;
AlertDialog.Footer = AlertDialogFooter;
AlertDialog.Title = AlertDialogTitle;
AlertDialog.Description = AlertDialogDescription;
AlertDialog.Action = AlertDialogAction;
AlertDialog.Cancel = AlertDialogCancel;

export { AlertDialog };
