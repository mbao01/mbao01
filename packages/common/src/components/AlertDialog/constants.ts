import { cva } from "../../libs";

export const getDialogCloseClasses = cva(
  "absolute right-4 top-4 rounded-xs opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-hidden focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
);
