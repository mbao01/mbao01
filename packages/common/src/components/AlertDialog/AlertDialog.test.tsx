import { describe, expect, it } from "vitest";
import { render, screen, within } from "@testing-library/react";
import { AlertDialog } from "./AlertDialog";
import userEvent from "@testing-library/user-event";

describe("AlertDialog", () => {
  const renderAlertDialog = () =>
    render(
      <AlertDialog>
        <AlertDialog.Trigger>Update profile</AlertDialog.Trigger>
        <AlertDialog.Content className="sm:max-w-[425px]">
          <AlertDialog.Header>
            <AlertDialog.Title>Update profile</AlertDialog.Title>
            <AlertDialog.Description>
              You are making changes to your profile here. Do you want to
              continue?
            </AlertDialog.Description>
          </AlertDialog.Header>
          <AlertDialog.Footer>
            <AlertDialog.Cancel variant="link">Cancel</AlertDialog.Cancel>
            <AlertDialog.Action>Continue</AlertDialog.Action>
          </AlertDialog.Footer>
        </AlertDialog.Content>
      </AlertDialog>
    );

  it("renders a an dialog trigger", () => {
    const { asFragment } = renderAlertDialog();

    expect(
      screen.getByRole("button", { name: "Update profile" })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a an alert dialog", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderAlertDialog();

    const dialogTrigger = screen.getByRole("button", {
      name: "Update profile",
    });

    expect(screen.queryByRole("alertdialog")).not.toBeInTheDocument();

    await user.click(dialogTrigger);

    const dialog = screen.getByRole("alertdialog");

    // dialog heading
    expect(
      within(dialog).getByRole("heading", { name: "Update profile" })
    ).toBeInTheDocument();

    // dialog description
    expect(
      within(dialog).getByText(
        "You are making changes to your profile here. Do you want to continue?"
      )
    ).toBeInTheDocument();

    // dialog footer
    expect(
      within(dialog).getByRole("button", { name: "Cancel" })
    ).toBeInTheDocument();
    expect(
      within(dialog).getByRole("button", { name: "Continue" })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should open and close alert dialog", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderAlertDialog();
    const dialogTrigger = screen.getByRole("button", {
      name: "Update profile",
    });

    expect(screen.queryByRole("alertdialog")).not.toBeInTheDocument();

    await user.click(dialogTrigger);

    expect(screen.queryByRole("alertdialog")).toBeInTheDocument();

    await user.click(screen.getByRole("button", { name: "Cancel" }));

    expect(screen.queryByRole("alertdialog")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should accept and close alert dialog", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderAlertDialog();
    const dialogTrigger = screen.getByRole("button", {
      name: "Update profile",
    });

    expect(screen.queryByRole("alertdialog")).not.toBeInTheDocument();

    await user.click(dialogTrigger);

    expect(screen.queryByRole("alertdialog")).toBeInTheDocument();

    await user.click(screen.getByRole("button", { name: "Continue" }));

    expect(screen.queryByRole("alertdialog")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
