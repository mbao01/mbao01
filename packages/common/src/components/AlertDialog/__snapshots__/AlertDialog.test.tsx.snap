// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlertDialog > renders a an alert dialog 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_3_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Update profile
  </button>
</DocumentFragment>
`;

exports[`AlertDialog > renders a an dialog trigger 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_0_"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Update profile
  </button>
</DocumentFragment>
`;

exports[`AlertDialog > should accept and close alert dialog 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_9_"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Update profile
  </button>
</DocumentFragment>
`;

exports[`AlertDialog > should open and close alert dialog 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_6_"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Update profile
  </button>
</DocumentFragment>
`;
